cmd/utils/cmd.go

```diff
@@ -293,6 +380,82 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
 	return nil
 }
 
+// ExportHistory exports blockchain history into the specified directory,
+// following the Era format.
+func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) error {
+	log.Info("Exporting blockchain history", "dir", dir)
+	if head := bc.CurrentBlock().Number.Uint64(); head < last {
+		log.Warn("Last block beyond head, setting last = head", "head", head, "last", last)
+		last = head
+	}
+	network := "unknown"
+	if name, ok := params.NetworkNames[bc.Config().ChainID.String()]; ok {
+		network = name
+	}
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return fmt.Errorf("error creating output directory: %w", err)
+	}
+	var (
+		start     = time.Now()
+		reported  = time.Now()
+		checksums []string
+	)
+	for i := uint64(first); i <= last; i += step {
+		gen := func() error {
+			var (
+				buf = bytes.NewBuffer(nil)
+				w   = era.NewBuilder(buf)
+			)
+			for j := uint64(0); j < step && j <= last-i; j++ {
+				nr := i + j
+				block := bc.GetBlockByNumber(nr)
+				if block == nil {
+					return fmt.Errorf("export failed on #%d: not found", nr)
+				}
+				receipts := bc.GetReceiptsByHash(block.Hash())
+				if receipts == nil {
+					return fmt.Errorf("export failed on #%d: receipts not found", nr)
+				}
+				td := bc.GetTd(block.Hash(), block.NumberU64())
+				if td == nil {
+					return fmt.Errorf("export failed on #%d: total difficulty not found", nr)
+				}
+				if err := w.Add(block, receipts, td); err != nil {
+					return err
+				}
+			}
+			root, err := w.Finalize()
+			if err != nil {
+				return fmt.Errorf("export failed to finalize %d: %w", step/i, err)
+			}
+
+			// Compute checksum of entire Era1.
+			checksums = append(checksums, common.Hash(sha256.Sum256(buf.Bytes())).Hex())
+
+			// Write Era1 to disk.
+			filename := path.Join(dir, era.Filename(network, int(i/step), root))
+			if err := os.WriteFile(filename, buf.Bytes(), os.ModePerm); err != nil {
+				return err
+			}
+
+			if time.Since(reported) >= 8*time.Second {
+				log.Info("Exporting blocks", "exported", i, "elapsed", common.PrettyDuration(time.Since(start)))
+				reported = time.Now()
+			}
+			return nil
+		}
+		if err := gen(); err != nil {
```

Comment:
Whats the advantage of making gen a function? Is it to parallelize it in the future?

Reply:
I would pull the 8 second logging out into the loop, maybe even get rid of the fun

Reply:
ahh yeah i used to have it so a bunch of `defer`s wouldn't pile up, but those are gone now so the whole function can go